scripts.py

```diff
@@ -1225,11 +1219,141 @@ def do_run(self):
+            #)
+
+        if identifiers:
+            if log:
+                log.info(
+                    'Restricted to %d specific identifiers.' % len(identifiers)
+                )
+            query = query.filter(
+                Edition.primary_identifier_id.in_([x.id for x in identifiers])
+            )
+        if identifier_type:
+            if log:
+                log.info(
+                    'Restricted to identifier type "%s".' % identifier_type
+                )
+            query = query.filter(Identifier.type==identifier_type)
+        '''
+        'SELECT editions.id AS editions_id, editions.data_source_id AS editions_data_source_id, editions.primary_identifier_id AS editions_primary_identifier_id, 
+        editions.title AS editions_title, editions.sort_title AS editions_sort_title, editions.subtitle AS editions_subtitle, editions.series AS editions_series, 
+        editions.series_position AS editions_series_position, editions.permanent_work_id AS editions_permanent_work_id, editions.author AS editions_author, 
+        editions.sort_author AS editions_sort_author, editions.language AS editions_language, editions.publisher AS editions_publisher, 
+        editions.imprint AS editions_imprint, editions.issued AS editions_issued, editions.published AS editions_published, editions.medium AS editions_medium, 
+        editions.cover_id AS editions_cover_id, editions.cover_full_url AS editions_cover_full_url, editions.cover_thumbnail_url AS editions_cover_thumbnail_url, 
+        editions.open_access_download_url AS editions_open_access_download_url, editions.simple_opds_entry AS editions_simple_opds_entry, 
+        editions.extra AS editions_extra \n
+        FROM editions, identifiers \n
+        WHERE editions.primary_identifier_id IN (:primary_identifier_id_1) AND identifiers.type = :type_1'
+        '''
+
+        if log:
+            log.info(
+                "Processing %d editions.", query.count()
+            )
+        return query.order_by(Edition.id)
+
+
+    def run(self, batch_size=10):
+        param_args = self.parse_command_line(self._db)
+        
+        #if param_args.identifiers:
+            # we're asked about a specific set of work contributors
+            #identifier_ids = [x.id for x in param_args.identifiers]
+
+        self.query = self.make_query(
+            self._db, param_args.identifier_type, param_args.identifiers, self.log
+        )
+
+        editions = True
+        offset = 0
+        while editions:
+            editions = self.query.offset(offset).limit(batch_size).all()
+            for edition in editions:
+                if edition.contributions:
+                    for contribution in edition.contributions:
+                        self.process_contribution(self._db, contribution, edition)
+            offset += batch_size
+            self._db.commit()
+        self._db.commit()
+
+
+    @classmethod
+    def process_contribution(cls, _db, contribution, edition):
```

Comment:
If you have a contribution you don't also need the edition; you can get it from `contribution.edition`.